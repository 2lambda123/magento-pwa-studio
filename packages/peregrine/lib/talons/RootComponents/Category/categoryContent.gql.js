import { gql } from '@apollo/client';

export const GET_PRODUCT_FILTERS_BY_CATEGORY = gql`
    query getProductFiltersByCategory(
        $categoryIdFilter: FilterEqualTypeInput!
    ) {
        products(filter: { category_id: $categoryIdFilter }) {
            aggregations {
                label
                count
                attribute_code
                options {
                    label
                    value
                }
            }
        }
    }
`;

export const GET_CATEGORY_CONTENT = gql`
    query getCategoryData($id: Int!) {
        category(id: $id) {
            id
            name
            description
        }
    }
`;

export default {
    getCategoryContentQuery: GET_CATEGORY_CONTENT,
    getProductFiltersByCategoryQuery: GET_PRODUCT_FILTERS_BY_CATEGORY
};
