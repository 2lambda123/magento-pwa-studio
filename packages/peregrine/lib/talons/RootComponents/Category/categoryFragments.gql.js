import { gql } from '@apollo/client';

export const CategoryFragment = gql`
    fragment CategoryFragment on CategoryTree {
        id
        description
        name
        product_count
        meta_title
        meta_keywords
        meta_description
    }
`;

export const ProductsFragment = gql`
    fragment ProductsFragment on Products {
        items {
            id
            name
            price {
                regularPrice {
                    amount {
                        currency
                        value
                    }
                }
            }
            small_image {
                url
            }
            url_key
            url_suffix
        }
        page_info {
            total_pages
        }
        total_count
    }
`;
