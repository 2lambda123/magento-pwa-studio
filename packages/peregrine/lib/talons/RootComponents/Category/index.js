export { useCategoryContent } from './useCategoryContent';
export { useNoProductsFound } from './useNoProductsFound';
export { useCategory } from './useCategory';
