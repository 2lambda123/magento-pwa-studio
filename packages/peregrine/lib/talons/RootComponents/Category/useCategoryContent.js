import { useCallback, useEffect, useState } from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { useAppContext } from '@magento/peregrine/lib/context/app';

const DRAWER_NAME = 'filter';
// TODO: This can be replaced by the value from `storeConfig when the PR,
// https://github.com/magento/graphql-ce/pull/650, is released.
const pageSize = 6;
const placeholderItems = Array.from({ length: pageSize }).fill(null);

/**
 * Returns props necessary to render the categoryContent component.
 *
 * @param {object} props.data - The results of a getCategory GraphQL query.
 *
 * @returns {object} result
 * @returns {number} result.categoryId - This category's ID.
 * @returns {string} result.categoryName - This category's name.
 * @returns {object} result.filters - The filters object.
 * @returns {func}   result.handleLoadFilters - A callback function to signal the user's intent to interact with the filters.
 * @returns {func}   result.handleOpenFilters - A callback function that actually opens the filter drawer.
 * @returns {object} result.items - The items in this category.
 * @returns {bool}   result.loadFilters - Whether or not the user has signalled their intent to interact with the filters.
 * @returns {string} result.pageTitle - The text to put in the browser tab for this page.
 */
export const useCategoryContent = props => {
    const {
        categoryId,
        data,
        queries: { getProductFiltersByCategory }
    } = props;

    const [loadFilters, setLoadFilters] = useState(false);
    const [, { toggleDrawer }] = useAppContext();

    const handleLoadFilters = useCallback(() => {
        setLoadFilters(true);
    }, [setLoadFilters]);
    const handleOpenFilters = useCallback(() => {
        setLoadFilters(true);
        toggleDrawer(DRAWER_NAME);
    }, [setLoadFilters, toggleDrawer]);

    const [getFilters, { data: filterData, error: filterError }] = useLazyQuery(
        getProductFiltersByCategory
    );

    useEffect(() => {
        if (filterError) {
            console.error(filterError);
        }
    }, [filterError]);

    useEffect(() => {
        if (categoryId) {
            getFilters({
                variables: {
                    categoryIdFilter: {
                        eq: categoryId
                    }
                }
            });
        }
    }, [categoryId, getFilters]);

    const filters = filterData ? filterData.products.aggregations : null;
    const items = data ? data.products.items : placeholderItems;
    const totalPagesFromData = data
        ? data.products.page_info.total_pages
        : null;
    const categoryName = data ? data.category.name : null;
    const categoryDescription = data ? data.category.description : null;
    // Note: STORE_NAME is injected by Webpack at build time.
    const pageTitle = categoryName
        ? `${categoryName} - ${STORE_NAME}`
        : STORE_NAME;

    return {
        categoryName,
        categoryDescription,
        filters,
        handleLoadFilters,
        handleOpenFilters,
        items,
        loadFilters,
        pageTitle,
        totalPagesFromData
    };
};
