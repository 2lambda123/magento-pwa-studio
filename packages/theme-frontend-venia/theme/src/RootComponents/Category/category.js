import { Component, createElement } from 'react';
import PropTypes from 'prop-types';

import classify from 'src/classify';
import Gallery from 'src/components/Gallery';
import mockData from './mockData';
import defaultClasses from './category.css';

const emptyData = [];

class Category extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            gallery: PropTypes.string,
            root: PropTypes.string,
            title: PropTypes.string
        })
    };

    state = {
        loaded: false
    };

    componentDidMount() {
        this.timer = window.setTimeout(() => {
            this.setState(() => ({ loaded: true }));
        }, 1000);
    }

    componentWillUnmount() {
        window.clearTimeout(this.timer);
    }

    render() {
        const { classes } = this.props;
        const data = this.state.loaded ? mockData : emptyData;

        return (
            <article className={classes.root}>
                <h1 className={classes.title}>
                    <span>Dresses</span>
                </h1>
                <section className={classes.gallery}>
                    <Gallery data={data} title="Dresses" />
                </section>
            </article>
        );
    }
}

export default classify(defaultClasses)(Category);
