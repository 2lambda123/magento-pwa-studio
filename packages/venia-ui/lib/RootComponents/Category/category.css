.root {
    padding: 1rem;
}

.title {
    font-size: 1.5rem;
    font-weight: 400;
    padding: 0 0.5rem;
}

.pagination {
    position: relative;
    bottom: 0;
}

.placeholder {
    height: 100vh;
}

.headerButtons {
    display: flex;
    justify-content: center;
    padding-bottom: 1.5rem;
}

.filterButton {
    margin-left: 0.5rem;
    display: inline-block;
    padding: 0.75rem 2rem;
    border: 1px solid black;
    border-radius: 3px;
    color: black;
    outline: none;
}

.categoryTitle {
    color: rgb(var(--venia-text));
    padding-bottom: 1rem;
    font-size: 1.375rem;
    font-weight: 300;
    line-height: 1.375rem;
    text-align: center;
}

.sortContainer {
    color: rgb(var(--venia-text-alt));
    text-align: center;
    font-size: 0.875rem;
    padding-bottom: 1rem;
}

.sortText {
    font-weight: 600;
}
