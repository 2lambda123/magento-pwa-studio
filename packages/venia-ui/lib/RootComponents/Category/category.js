import React, { useEffect, useMemo, useRef, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { number, shape, string } from 'prop-types';
import { useLazyQuery, useQuery } from '@apollo/react-hooks';
import { usePagination } from '@magento/peregrine';

import { mergeClasses } from '../../classify';
import { fullPageLoadingIndicator } from '../../components/LoadingIndicator';

import CategoryContent from './categoryContent';
import defaultClasses from './category.css';
import { Meta } from '../../components/Head';
import {
    getFiltersFromSearch,
    getFilterInput
} from '@magento/peregrine/lib/talons/FilterModal/helpers';

import GET_CATEGORY from '../../queries/getCategory.graphql';
import FILTER_INTROSPECTION from '../../queries/introspection/filterIntrospectionQuery.graphql';

const Category = props => {
    const { id, pageSize } = props;
    const classes = mergeClasses(defaultClasses, props.classes);
    const [paginationValues, paginationApi] = usePagination();
    const { currentPage, totalPages } = paginationValues;
    const { setCurrentPage, setTotalPages } = paginationApi;

    const [sort, setSort] = useState({
        sortAttribute: 'relevance',
        sortDirection: 'DESC'
    });
    const { sortAttribute, sortDirection } = sort;
    const sortControl = {
        currentSort: sort,
        setSort: setSort
    };

    const pageControl = {
        currentPage,
        setPage: setCurrentPage,
        totalPages
    };

    const [runQuery, queryResponse] = useLazyQuery(GET_CATEGORY);
    const { loading, error, data } = queryResponse;
    const { search } = useLocation();

    // Keep track of the search terms so we can tell when they change.
    const previousSearch = useRef(search);

    // Get "allowed" filters by intersection of schema and aggregations
    const { data: introspectionData, error: introspectionError } = useQuery(
        FILTER_INTROSPECTION
    );

    useEffect(() => {
        if (introspectionError) {
            console.error(introspectionError);
        }
    }, [introspectionError]);

    // Create a type map we can reference later to ensure we pass valid args
    // to the graphql query.
    // For example: { category_id: 'FilterEqualTypeInput', price: 'FilterRangeTypeInput' }
    const filterTypeMap = useMemo(() => {
        const typeMap = new Map();
        if (introspectionData) {
            introspectionData.__type.inputFields.forEach(({ name, type }) => {
                typeMap.set(name, type.name);
            });
        }
        return typeMap;
    }, [introspectionData]);

    // Run the category query immediately and whenever its variable values change.
    useEffect(() => {
        // Wait until we have the type map to fetch product data.
        if (!filterTypeMap.size) {
            return;
        }

        const filters = getFiltersFromSearch(search);

        // Construct the filter arg object.
        const newFilters = {};
        filters.forEach((values, key) => {
            newFilters[key] = getFilterInput(values, filterTypeMap.get(key));
        });

        // Use the category id for the current category page regardless of the
        // applied filters. Follow-up in PWA-404.
        newFilters['category_id'] = { eq: String(id) };

        runQuery({
            variables: {
                currentPage: Number(currentPage),
                id: Number(id),
                filters: newFilters,
                onServer: false,
                pageSize: Number(pageSize),
                sort: { [String(sortAttribute)]: String(sortDirection) }
            }
        });
        window.scrollTo({
            left: 0,
            top: 0,
            behavior: 'smooth'
        });
    }, [
        currentPage,
        filterTypeMap,
        id,
        pageSize,
        runQuery,
        search,
        sortAttribute,
        sortDirection
    ]);

    const totalPagesFromData = data
        ? data.products.page_info.total_pages
        : null;

    useEffect(() => {
        setTotalPages(totalPagesFromData);
        return () => {
            setTotalPages(null);
        };
    }, [setTotalPages, totalPagesFromData]);

    // If we get an error after loading we should try to reset to page 1.
    // If we continue to have errors after that, render an error message.
    useEffect(() => {
        if (error && !loading && currentPage !== 1) {
            setCurrentPage(1);
        }
    }, [currentPage, error, loading, setCurrentPage]);

    // Reset the current page back to one (1) when the search string or filters
    // change.
    useEffect(() => {
        // We don't want to compare page value.
        const prevSearch = new URLSearchParams(previousSearch.current);
        const nextSearch = new URLSearchParams(search);
        prevSearch.delete('page');
        nextSearch.delete('page');

        if (prevSearch.toString() != nextSearch.toString()) {
            // The search term changed.
            setCurrentPage(1);
            // And update the ref.
            previousSearch.current = search;
        }
    }, [previousSearch, search, setCurrentPage]);

    if (error && currentPage === 1 && !loading) {
        if (process.env.NODE_ENV !== 'production') {
            console.error(error);
        }
        return <div>Data Fetch Error</div>;
    }

    // Show the loading indicator until data has been fetched.
    if (totalPagesFromData === null) {
        return fullPageLoadingIndicator;
    }

    return (
        <>
            <Meta
                name="description"
                content={
                    data && data.category && data.category.meta_description
                }
            />
            <CategoryContent
                categoryId={id}
                classes={classes}
                data={loading ? null : data}
                pageControl={pageControl}
                sortControl={sortControl}
            />
        </>
    );
};

Category.propTypes = {
    classes: shape({
        gallery: string,
        root: string,
        title: string
    }),
    id: number,
    pageSize: number
};

Category.defaultProps = {
    id: 3,
    // TODO: This can be replaced by the value from `storeConfig when the PR,
    // https://github.com/magento/graphql-ce/pull/650, is released.
    pageSize: 6
};

export default Category;
