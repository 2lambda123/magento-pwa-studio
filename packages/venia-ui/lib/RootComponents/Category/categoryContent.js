import React, { Fragment, Suspense } from 'react';
import { array, shape, string } from 'prop-types';
import RichContent from '../../components/RichContent';

import { useCategoryContent } from '@magento/peregrine/lib/talons/RootComponents/Category';

import NoProductsFound from './NoProductsFound';
import { mergeClasses } from '../../classify';
import { Title } from '../../components/Head';
import Breadcrumbs from '../../components/Breadcrumbs';
import Gallery from '../../components/Gallery';
import ProductSort from '../../components/ProductSort';
import Pagination from '../../components/Pagination';
import defaultClasses from './category.css';

const FilterModal = React.lazy(() => import('../../components/FilterModal'));
import GET_PRODUCT_FILTERS_BY_CATEGORY from '../../queries/getProductFiltersByCategory.graphql';

const CategoryContent = props => {
    const { categoryId, data, pageControl, sortProps } = props;
    const [currentSort] = sortProps;

    const talonProps = useCategoryContent({
        categoryId,
        data,
        queries: {
            getProductFiltersByCategory: GET_PRODUCT_FILTERS_BY_CATEGORY
        }
    });

    const {
        categoryName,
        categoryDescription,
        filters,
        handleLoadFilters,
        handleOpenFilters,
        items,
        pageTitle,
        totalPagesFromData
    } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);

    const header = filters ? (
        <Fragment>
            <div className={classes.headerButtons}>
                <button
                    className={classes.filterButton}
                    onClick={handleOpenFilters}
                    onFocus={handleLoadFilters}
                    onMouseOver={handleLoadFilters}
                    type="button"
                >
                    {'Filter'}
                </button>
                <ProductSort sortProps={sortProps} />
            </div>
            <div className={classes.sortContainer}>
                {'Items sorted by '}
                <span className={classes.sortText}>{currentSort.sortText}</span>
            </div>
        </Fragment>
    ) : null;

    // If you want to defer the loading of the FilterModal until user interaction
    // (hover, focus, click), simply add the talon's `loadFilters` prop as
    // part of the conditional here.
    const modal = filters ? <FilterModal filters={filters} /> : null;

    const categoryDescriptionElement = categoryDescription ? (
        <RichContent html={categoryDescription} />
    ) : null;

    const content =
        totalPagesFromData === 0 ? (
            <NoProductsFound categoryId={categoryId} />
        ) : (
            <Fragment>
                <section className={classes.gallery}>
                    <Gallery items={items} />
                </section>
                <div className={classes.pagination}>
                    <Pagination pageControl={pageControl} />
                </div>
            </Fragment>
        );

    return (
        <Fragment>
            <Breadcrumbs categoryId={categoryId} />
            <Title>{pageTitle}</Title>
            <article className={classes.root}>
                <h1 className={classes.title}>
                    <div className={classes.categoryTitle}>{categoryName}</div>
                </h1>
                {categoryDescriptionElement}
                {header}
                {content}
                <Suspense fallback={null}>{modal}</Suspense>
            </article>
        </Fragment>
    );
};

export default CategoryContent;

CategoryContent.propTypes = {
    classes: shape({
        filterContainer: string,
        gallery: string,
        headerButtons: string,
        pagination: string,
        root: string,
        title: string
    }),
    // sortProps contains the following structure:
    // [{sortDirection: string, sortAttribute: string, sortText: string},
    // React.Dispatch<React.SetStateAction<{sortDirection: string, sortAttribute: string, sortText: string}]
    sortProps: array
};
