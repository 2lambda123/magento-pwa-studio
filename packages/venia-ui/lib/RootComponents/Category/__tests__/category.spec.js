import React from 'react';
import { act } from 'react-test-renderer';
import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';
import Category from '../category';

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({ push: jest.fn() })),
    useLocation: jest.fn(() => ({ pathname: '', search: '' }))
}));

jest.mock('../../../components/Head', () => ({
    HeadProvider: ({ children }) => <div>{children}</div>,
    Title: () => 'Title',
    Meta: () => 'Meta'
}));

jest.mock('../categoryContent', () => 'CategoryContent');

const mockUseSort = jest
    .fn()
    .mockReturnValue([
        {
            sortText: 'Best Match',
            sortAttribute: 'relevance',
            sortDirection: 'DESC'
        },
        jest.fn()
    ])
    .mockName('mockUseSort');

const mockSetCurrentPage = jest.fn().mockName('mockSetCurrentPage');

jest.mock('@magento/peregrine', () => {
    const usePagination = jest.fn(() => [
        {
            currentPage: 3,
            totalPages: 6
        },
        {
            setCurrentPage: jest
                .fn()
                .mockImplementation(() => mockSetCurrentPage()),
            setTotalPages: jest.fn()
        }
    ]);
    const useSort = jest.fn().mockImplementation(() => mockUseSort());

    return {
        useSort,
        usePagination
    };
});

jest.mock('@apollo/react-hooks', () => {
    const useQuery = jest.fn().mockReturnValue({
        data: {
            __type: {
                inputFields: []
            }
        },
        error: null,
        loading: false
    });
    const runQuery = jest.fn();
    const queryResult = {
        data: {
            products: {
                page_info: {
                    total_pages: 6
                }
            }
        },
        error: null,
        loading: false
    };
    const useLazyQuery = jest.fn(() => [runQuery, queryResult]);

    return { useLazyQuery, useQuery };
});

const categoryProps = {
    id: 3,
    pageSize: 6
};
const tree = createTestInstance(<Category {...categoryProps} />);

test('renders the correct tree', () => {
    expect(tree.toJSON()).toMatchSnapshot();
});

const testCases = [
    [
        'sortText does not reset',
        {
            sortText: 'Changed',
            sortAttribute: 'relevance',
            sortDirection: 'DESC'
        },
        0
    ],
    [
        'sortAttribute resets',
        {
            sortText: 'Best Match',
            sortAttribute: 'Changed',
            sortDirection: 'DESC'
        },
        1
    ],
    [
        'sortDirection resets',
        {
            sortText: 'Best Match',
            sortAttribute: 'relevance',
            sortDirection: 'Changed'
        },
        1
    ]
];

test.each(testCases)(
    'Changing %s current page to 1.',
    (description, sortParams, expected) => {
        mockUseSort.mockReturnValueOnce([sortParams, jest.fn()]);
        act(() => {
            tree.update(<Category {...categoryProps} />);
        });

        expect(mockSetCurrentPage).toHaveBeenCalledTimes(expected);
    }
);
