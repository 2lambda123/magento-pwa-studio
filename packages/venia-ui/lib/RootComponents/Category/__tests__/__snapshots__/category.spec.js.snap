// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
Array [
  "Meta",
  <CategoryContent
    categoryId={3}
    classes={Object {}}
    data={
      Object {
        "products": Object {
          "page_info": Object {
            "total_pages": 6,
          },
        },
      }
    }
    pageControl={
      Object {
        "currentPage": 3,
        "setPage": [MockFunction],
        "totalPages": 6,
      }
    }
    sortProps={
      Array [
        Object {
          "sortAttribute": "relevance",
          "sortDirection": "DESC",
          "sortText": "Best Match",
        },
        [MockFunction],
      ]
    }
  />,
]
`;
